### PR TITLE
IF: Don't drop late blocks

### LINKED ISSUE #2159: IF: Don't drop late blocks
Currently, any blocks received while producing a block are dropped and not evaluated. IF includes QC info in block extensions. Do not drop blocks, but rather incorporate their QC info into the `fork_database`. Do not automatically abort block but incorporate the block into the `fork_database` for use in best fork evaluation. If LIB advances such that current block being produced is no longer built on LIB, abort the block immediate, re-evaluate best fork, and build block off the new best fork.

Related to #2125

### PR DESCRIPTION
**Pull request by heifner**
- Instead of dropping blocks when producing, incorporate blocks received while producing into the fork database.
  - Potentially fork switch to newly incorporated blocks before producing a block
  - Use block id instead of block number for irreversible so that correct fork branch is made irreversible

Resolves #2159

### CODE REVIEW

File: libraries/chain/controller.cpp [hunk 1]
```diff
@@ -1212,23 +1218,28 @@ struct controller_impl {
...
          return;
 
       auto mark_branch_irreversible = [&, this](auto& forkdb) {
-         auto branch = forkdb.fetch_branch( fork_db_head(forkdb, irreversible_mode())->id(), new_lib );
+         auto branch = (if_lib > 0) ? forkdb.fetch_branch( if_irreversible_block_id, new_lib)
+                                    : forkdb.fetch_branch( fork_db_head(forkdb, irreversible_mode())->id(), new_lib );
          try {
+            auto should_process = [&](auto& bsp) {
+               return read_mode == db_read_mode::IRREVERSIBLE || bsp->is_valid();
+            };
```
Comment (greg7mdp) [inline libraries/chain/controller.cpp:1233]: Why is this necessary? I thought that lib and its ancestors are guaranteed to be valid.

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -1212,23 +1218,28 @@ struct controller_impl {
...
-      const uint32_t if_lib = if_irreversible_block_num;
+      uint32_t if_lib = block_header::num_from_id(if_irreversible_block_id);
       const uint32_t new_lib = if_lib > 0 ? if_lib : fork_db_head_irreversible_blocknum();
 
       if( new_lib <= lib_num )
          return;
 
       auto mark_branch_irreversible = [&, this](auto& forkdb) {
-         auto branch = forkdb.fetch_branch( fork_db_head(forkdb, irreversible_mode())->id(), new_lib );
+         auto branch = (if_lib > 0) ? forkdb.fetch_branch( if_irreversible_block_id, new_lib)
```
Comment (greg7mdp) [inline libraries/chain/controller.cpp:1228]: That's always going to be a single block, right, since `new_lib == num_from_id(if_irreversible_block_id);`

File: libraries/chain/controller.cpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/controller.cpp:1233]: No, they are not guaranteed to be valid. We are putting non-validated blocks into the fork database. A different branch than our best head can become irreversible which means we have not necessarily validated them. The `lib_advance_test.py` actually triggered this case.

An alternative more complex implementation would be to do a fork switch so we can move irreversible all the way at this point. Instead this moves irreversible as much as possible right here and then waits for a `maybe_switch_forks` call to happen *soon* to switch to the new branch and then move irreversible after that. *soon* in this case should be before the next produced block or on the next received block.

I'll add a comment that explains this.

File: libraries/chain/controller.cpp (hunk 2 above)
Comment (heifner) [reply libraries/chain/controller.cpp:1228]: No, this returns all blocks from root up to `new_lib` on branch with `if_irreversible_block_id`.

Comment (greg7mdp) [reply libraries/chain/controller.cpp:1228]: Aren't they the same though?

File: libraries/chain/controller.cpp (hunk 1 above)
Comment (greg7mdp) [reply libraries/chain/controller.cpp]: Even if we have not validated the block, it became irreversible because we received a strong QC making it irreversible, so I am not sure what benefit there is to not accept it as irreversible. Even if it did not validate, I don't know what we can do since a quorum of finalizers already accepted it.

File: libraries/chain/controller.cpp (hunk 2 above)
Comment (greg7mdp) [reply libraries/chain/controller.cpp:1228]: Looking at the code I see:
```
if_lib = block_header::num_from_id(if_irreversible_block_id);
new_lib = if_lib;
```

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: INTERNALS
summary: Incorporate blocks received while producing into fork database. Previously late blocks were dropped. 
Note:end

File: libraries/chain/controller.cpp (hunk 2 above)
Comment (heifner) [reply libraries/chain/controller.cpp:1228]: `new_lib` does equal `num_from_id(if_irreversible_block_id)`

However, `forkdb.fetch_branch( if_irreversible_block_id, new_lib)` returns all blocks from the `root` of the fork database up to and including `if_irreversible_block_id` trimming off any after `new_lib` which there will not be any. Could not pass in `new_lib` as it doesn't matter. However, might as well pass it in as it makes it same as pre-instant-finality.

Here is an example from a unittest where LIB moved by two blocks. You can see the branch is of size 2 in this case.

image

File: libraries/chain/controller.cpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/controller.cpp]: It is our job as a validating node to validate blocks. If some disaster scenario happens we want to be in a good state. Also our block log should not contain data that is not in our state. We need a consistent state and block log in case we have to shutdown. Not even sure what kind of strange corner cases there would be for a node that has blocks in the block log that are not incorporated into the state.

Comment (heifner) [reply libraries/chain/controller.cpp]: Yes, it would be split. But who is to say this isn't the correct chain. Manual intervention is needed.

File: libraries/chain/controller.cpp (hunk 2 above)
Comment (greg7mdp) [reply libraries/chain/controller.cpp:1228]: Got it, sorry, for some reason I keep misunderstanding that `fetch_branch` API.

File: libraries/chain/controller.cpp (hunk 1 above)
Comment (greg7mdp) [reply libraries/chain/controller.cpp]: Do you agree that the node might as well terminate in that case?

Comment (greg7mdp) [reply libraries/chain/controller.cpp]: @heifner I modified your comment by mistake. Sorry!

Comment (greg7mdp) [reply libraries/chain/controller.cpp]: > But who is to say this isn't the correct chain.

A quorum of finalizers said so.

Comment (heifner) [reply libraries/chain/controller.cpp]: > @heifner I modified your comment by mistake. Sorry!

Done that many times myself. No worries.

Comment (greg7mdp) [reply libraries/chain/controller.cpp]: OK, I guess I understand the logic here. We expect that, even if the block made irreversible by the received QC has not been validated yet, we will validate it eventually and then be able to commit it to the state. Seems like a weird state to be in though.

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -816,7 +816,7 @@ struct controller_impl {
    block_log                       blog;
    std::optional<pending_state>    pending;
    fork_database                   fork_db;
-   std::atomic<uint32_t>           if_irreversible_block_num{0};
+   block_id_type                   if_irreversible_block_id;
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/controller.cpp:819]: Correct, only accessed by the main thread now.

File: libraries/chain/controller.cpp [hunk 4]
```diff
@@ -3422,7 +3462,6 @@ struct controller_impl {
          trusted_producer_light_validation = old_value;
       });
       try {
-         EOS_ASSERT( bsp, block_validate_exception, "null block" );
```
[the remark this answers is not in the document]
Comment (heifner) [reply libraries/chain/controller.cpp:3425]: It shouldn't as this is more of a true `assert` then an `EOS_ASSERT`.

File: plugins/producer_plugin/producer_plugin.cpp [hunk 5]
```diff
@@ -667,38 +667,46 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
...
       }
 
+      if (in_producing_mode()) {
+         fc_ilog(_log, "producing, incoming block #${num} id: ${id}", ("num", blk_num)("id", id));
+         const block_handle& bh = obt ? *obt : btf.get();
+         chain.accept_block(bh);
+         return true; // return true because block was accepted
+      }
+
+      if (now - block->timestamp < fc::minutes(5) || (blk_num % 1000 == 0)) // only log every 1000 during sync
```
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/producer_plugin/producer_plugin.cpp:707]: Sure, probably makes sense

**Review by linh2931 — APPROVED** [review]

**Review by greg7mdp — APPROVED** [review]
